File: src/format.js
/**
 * SPDX-FileCopyrightText: 2026 Nissaar
 * SPDX-License-Identifier: AGPL-3.0-or-later
 */

import { getCanonicalLocale, translate as t } from '@nextcloud/l10n'

/**
 * "2024-07" as a month a person would read.
 *
 * @param {string} yearMonth a month key like "2024-07"
 * @return {string}
 */
export function monthLabel(yearMonth) {
	const [year, month] = String(yearMonth).split('-').map(Number)
	if (!year || !month) {
		return yearMonth
	}
	return new Date(Date.UTC(year, month - 1, 1)).toLocaleDateString(getCanonicalLocale(), {
		month: 'long',
		year: 'numeric',
		timeZone: 'UTC',
	})
}

/**
 * @param {number} seconds epoch seconds
 * @return {string}
 */
export function dateLabel(seconds) {
	if (!seconds) {
		return ''
	}
	return new Date(seconds * 1000).toLocaleDateString(getCanonicalLocale(), {
		day: 'numeric',
		month: 'short',
		year: 'numeric',
	})
}

/**
 * @param {number} bytes a size in bytes
 * @return {string}
 */
export function sizeLabel(bytes) {
	if (!bytes) {
		return ''
	}
	const units = ['B', 'KB', 'MB', 'GB', 'TB']
	let value = bytes
	let unit = 0
	while (value >= 1024 && unit < units.length - 1) {
		value /= 1024
		unit++
	}
	return `${value < 10 && unit > 0 ? value.toFixed(1) : Math.round(value)} ${units[unit]}`
}

/**
 * Explains where a photo's date came from, for the cases where it looks wrong.
 *
 * @param {string} source one of the DateSource values
 * @return {string}
 */
export function dateSourceLabel(source) {
	switch (source) {
		case 'exif':
			return t('photocleaner', 'Date taken, from the photo itself')
		case 'filename':
			return t('photocleaner', 'Date read from the file name')
		case 'upload':
			return t('photocleaner', 'Date this file reached the server')
		default:
			return t('photocleaner', 'Date the file was last changed — no capture date was available')
	}
}

File: android/app/src/main/java/xyz/photosweep/ui/Format.kt
package xyz.photosweep.ui

import java.time.Instant
import java.time.YearMonth
import java.time.ZoneId
import java.time.format.DateTimeFormatter
import java.util.Locale

/*
 * Built per call rather than held in a field. `Locale.getDefault()` is read once when
 * a field is initialised, so a formatter cached here would keep rendering month names
 * in whatever language the app started in, even after the user changes the system
 * language and comes back.
 */
private fun monthFormatter() = DateTimeFormatter.ofPattern("LLLL yyyy", Locale.getDefault())

private fun dayFormatter() = DateTimeFormatter.ofPattern("d MMM yyyy", Locale.getDefault())

/** "2024-07" as a month a person would read. */
fun monthLabel(yearMonth: String): String = try {
    YearMonth.parse(yearMonth).format(monthFormatter())
} catch (e: Exception) {
    yearMonth
}

fun dateLabel(epochSeconds: Long, zone: ZoneId = ZoneId.systemDefault()): String =
    if (epochSeconds <= 0L) {
        ""
    } else {
        Instant.ofEpochSecond(epochSeconds).atZone(zone).format(dayFormatter())
    }

fun sizeLabel(bytes: Long): String {
    if (bytes <= 0) return ""
    val units = listOf("B", "KB", "MB", "GB", "TB")
    var value = bytes.toDouble()
    var unit = 0
    while (value >= 1024 && unit < units.lastIndex) {
        value /= 1024
        unit++
    }
    return if (value < 10 && unit > 0) {
        String.format(Locale.getDefault(), "%.1f %s", value, units[unit])
    } else {
        String.format(Locale.getDefault(), "%.0f %s", value, units[unit])
    }
}

/**
 * Explains where a photo's date came from.
 *
 * Shown because a photo landing in an unexpected month is confusing right up until you
 * can see that the app only had a modification time to go on.
 */
fun dateSourceLabel(source: String): String = when (source) {
    "exif" -> "Date taken, from the photo itself"
    "filename" -> "Date read from the file name"
    "upload" -> "Date this file reached the server"
    else -> "Date the file was last changed — no capture date was available"
}
